vxquery-core/src/main/java/org/apache/vxquery/jsonparser/JSONParser.java

```diff
@@ -0,0 +1,188 @@
+                    abStack.get(abStack.size() - 1).addItem(atomic);
+                } else if (lastItem == "object") {
+                    obStack.get(obStack.size() - 1).addItem(spStack.get(spStack.size() - 1), atomic);
+                }
+            }
+            if (token == JsonToken.VALUE_STRING) {
+                atomic.reset();
+                out.write(ValueTag.XS_STRING_TAG);
+                svb.write(parser.getText(), out);
+                if (lastItem == "array") {
+                    abStack.get(abStack.size() - 1).addItem(atomic);
+                } else if (lastItem == "object") {
+                    obStack.get(obStack.size() - 1).addItem(spStack.get(spStack.size() - 1), atomic);
+                }
+            }
+            if (token == JsonToken.VALUE_NUMBER_FLOAT) {
+                atomic.reset();
+                out.write(ValueTag.XS_DOUBLE_TAG);
+                out.writeDouble(parser.getDoubleValue());
+                if (lastItem == "array") {
+                    abStack.get(abStack.size() - 1).addItem(atomic);
+                } else if (lastItem == "object") {
+                    obStack.get(obStack.size() - 1).addItem(spStack.get(spStack.size() - 1), atomic);
+                }
+            }
+            if (token == JsonToken.END_ARRAY) {
+                if (itemStack.size() == 1) {
+                    abStack.get(abStack.size() - 1).finish();
+                    abStack.remove(abStack.size() - 1);
+                    itemStack.remove(itemStack.size() - 1);
+                    nested = true;
+                    lastItem = "array";
+                } else if (itemStack.get(itemStack.size() - 2) == "array") {
+                    abStack.get(abStack.size() - 1).finish();
+                    abStack.remove(abStack.size() - 1);
+                    abStack.get(abStack.size() - 1).addItem(abvsStack.get(abvsStack.size() - 1));
+                    abvsStack.remove(abvsStack.size() - 1);
+                    itemStack.remove(itemStack.size() - 1);
+                    lastItem = "array";
+                } else if (itemStack.get(itemStack.size() - 2) == "object") {
+                    abStack.get(abStack.size() - 1).finish();
+                    abStack.remove(abStack.size() - 1);
+                    obStack.get(obStack.size() - 1).addItem(spStack.get(spStack.size() - 1),
+                            abvsStack.get(abvsStack.size() - 1));
+                    abvsStack.remove(abvsStack.size() - 1);
+                    itemStack.remove(itemStack.size() - 1);
+                    nested = true;
+                    lastItem = "object";
+                }
+
+                startItem = "array";
+
+            }
+            if (token == JsonToken.END_OBJECT) {
+                if (itemStack.size() == 1) {
+                    obStack.get(obStack.size() - 1).finish();
+                    obStack.remove(obStack.size() - 1);
+                    spStack.remove(spStack.size() - 1);
```

Comment:
There's quite a bit of redundancy in these cases (also for the END_ARRAY case), e.g. the first 3 lines. It would be nice to extract that and have the code only once.